funsor/distributions.py

```diff
@@ -451,22 +452,77 @@ def eager_mvn(loc, scale_tril, value):
 
 
 # Create a Gaussian from a ground observation.
-@eager.register(MultivariateNormal, Tensor, Tensor, Variable)
-@eager.register(MultivariateNormal, Variable, Tensor, Tensor)
+# TODO refactor this logic into Gaussian.eager_subs() and
+#   here return Gaussian(...scale_tril...)(value=loc-value).
+@eager.register(MultivariateNormal, (Variable, Contraction), Tensor, (Variable, Contraction))
+@eager.register(MultivariateNormal, (Variable, Contraction), Tensor, Tensor)
+@eager.register(MultivariateNormal, Tensor, Tensor, (Variable, Contraction))
 def eager_mvn(loc, scale_tril, value):
-    if isinstance(loc, Variable):
-        loc, value = value, loc
+    affine = loc - value
+    if not isinstance(affine, Contraction):
+        return None  # lazy
+
+    const, coeffs = extract_affine(affine)
+    if not isinstance(const, Tensor):
+        return None  # lazy
+    if not all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values()):
+        return None  # lazy
+
+    # Dovetail to avoid variable name collision in einsum.
+    equations1 = [''.join(c if c in '.,->' else chr(ord(c) * 2))
+                  for _, eqn in coeffs.values() for c in eqn]
+    equations2 = [''.join(c if c in '.,->' else chr(ord(c) * 2 + 1))
+                  for _, eqn in coeffs.values() for c in eqn]
 
-    dim, = loc.output.shape
-    inputs, (loc, scale_tril) = align_tensors(loc, scale_tril)
-    inputs.update(value.inputs)
-    int_inputs = OrderedDict((k, v) for k, v in inputs.items() if v.dtype != 'real')
+    real_inputs = OrderedDict((k, v) for k, v in affine.inputs.items() if v.dtype == 'real')
+    assert len(real_inputs) == len(coeffs)
+    tensors = [coeffs[k] for k in real_inputs] + [const, scale_tril]
+    inputs, tensors = align_tensors(*tensors)
 
-    precision = cholesky_inverse(scale_tril)
+    # Incorporate scale_tril before broadcasting.
+    scale_tril = tensors[-1]
+    prec_sqrt = cholesky_inverse(scale_tril)  # FIXME is transpose correct?
+    for i in range(len(coeffs)):
+        coeff = tensors[i]
+        inputs1, output1 = equations1[i].split('->')
+        input11, _ = inputs1.split(',')
+        inputs2, output2 = equations2[i].split('->')
+        input21, _ = inputs2.split(',')
+        coeff = torch.einsum(f'...{input11},...{output1}{output2}->...{input21}',
+                             coeff, prec_sqrt)
+        tensors[i] = coeff
+
+    tensors = torch.broadcast_tensors(*tensors)
+    coeffs, const, scale_tril = tensors[:-2], tensors[-2], tensors[-1]
+    batch_shape = const.shape
+    dim = sum(d.num_elements for d in real_inputs.values())
+
+    loc = BlockVector(batch_shape + (dim,))
+    loc[..., 0] = -const / coeffs[0]  # FIXME consider directly constructing info_vec
```

Comment:
I think this should be fine. Constructing info_vec directly seems complicated to me.

Reply:
@fehiepsi can you help me out here, I copied this from the univariate case and could use help with the linear algebra to incorporate `const`.

Reply:
Rethinking about this, I do agree that it is better to compute info_vec directly. The way we do this in the univariate case might not be suitable for the case `coeffs[0] = 0`. I think we can process as
```
(a + xB + yC)Psqrt @ transpose_of_previous_term
= (a @ Psqrt + xBPsqrt + yCPsqrt) @ transpose...
= (a' + xB' + yC') @ transpose...
Hence, precision = (B' C') @ (B' C').T
and info_vec = (a' @ B'.T, a' @ C'.T)
```

Reply:
I think `loc @ info_vec` is simply `a' @ a'.T`...

Reply:
Great, I've pushed attempted fixes. Can you take another look?